tests/integration/pilot/tunneling/tunneling_test.go

```diff
@@ -0,0 +1,416 @@
+
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		common.ApplyFileOrFail(ctx, meshNs.Name(), path.Join(env.IstioSrc, "samples/sleep/sleep.yaml"))
+		waitForPodsReadyOrFail(ctx, meshNs.Name(), "sleep")
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		templateParams := map[string]interface{}{
+			"codecType":  proxyHTTPVersion,
+			"tlsEnabled": proxyTLSEnabled,
+			"protocol":   serviceEntryProtocol(proxyHTTPVersion, proxyTLSEnabled),
+		}
+		ctx.ConfigIstio().EvalFile(meshNs.Name(), templateParams, "forward-proxy/service-entry.tmpl.yaml").ApplyOrFail(ctx)
+		ctx.ConfigIstio().EvalFile(meshNs.Name(), templateParams, "forward-proxy/destination-rule.tmpl.yaml").ApplyOrFail(ctx)
+		common.ApplyFileOrFail(ctx, externalNs.Name(), "forward-proxy/ssl-certificate-configmap.yaml")
+		common.ApplyFileOrFail(ctx, externalNs.Name(), "forward-proxy/ssl-private-key-configmap.yaml")
+		ctx.ConfigIstio().EvalFile(externalNs.Name(), templateParams, "forward-proxy/configmap.tmpl.yaml").ApplyOrFail(ctx)
+		common.ApplyFileOrFail(ctx, externalNs.Name(), "forward-proxy/deployment.yaml")
+		waitForPodsReadyOrFail(ctx, externalNs.Name(), "external-forward-proxy")
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		common.ApplyFileOrFail(ctx, meshNs.Name(), "external-app/service-entry.yaml")
+		common.ApplyFileOrFail(ctx, externalNs.Name(), "external-app/ssl-certificate-configmap.yaml")
+		common.ApplyFileOrFail(ctx, externalNs.Name(), "external-app/ssl-private-key-configmap.yaml")
+		common.ApplyFileOrFail(ctx, externalNs.Name(), "external-app/configmap.yaml")
+		common.ApplyFileOrFail(ctx, externalNs.Name(), "external-app/deployment.yaml")
+		waitForPodsReadyOrFail(ctx, externalNs.Name(), "external-app")
+	}()
+
+	wg.Wait()
+	makeExternalServicesResolvable(ctx, externalNs.Name(), meshNs.Name())
+
+	for _, tc := range testCases() {
+		for _, res := range tc.istioResourcesToApply {
+			common.ApplyFileOrFail(ctx, meshNs.Name(), res)
+		}
+
+		for _, protocol := range tc.protocols {
+			ctx.NewSubTest(fmt.Sprintf("%s/%s-request", tc.name, protocol)).Run(func(ctx framework.TestContext) {
+				// requests will fail until istio-proxy gets the Envoy configuration from istiod, so retries are necessary
+				retry.UntilSuccessOrFail(ctx, func() error {
+					if err := executeRequestToExternalApp(ctx, meshNs.Name(), protocol, tc.name); err != nil {
+						return err
+					}
+					externalForwardProxyIP := getPodIP(ctx, externalNs.Name(), "external-forward-proxy")
+					if err := verifyThatRequestWasTunneled(ctx, externalNs.Name(), externalForwardProxyIP, protocol, tc.name); err != nil {
+						return err
+					}
+					return nil
+				}, retry.Timeout(10*time.Second), retry.Delay(1*time.Second))
+			})
+		}
+
+		for _, res := range tc.istioResourcesToApply {
+			common.DeleteFileOrFail(ctx, meshNs.Name(), res)
+		}
+		// make sure that configuration changes were pushed to sidecar proxies
+		time.Sleep(10 * time.Second)
```

Comment:
Sleep in test is not great: https://github.com/istio/istio/wiki/Test-Flakes#timing-dependencies

Reply:
Yes, I definitely agree with you and I am not proud of this approach, but this sleep is necessary to ensure that the updated configuration was pushed to sidecar proxies. I wasn't sure if I can use sleep in integration tests, but I searched `time.Sleep` in other tests and I saw that it's sometimes used, so I thought that it's acceptable. To avoid sleep I would have to get config dumps from sidecars and check if were updated. It might be quite complicated, but I can try to do that.

Reply:
The preferred approach is to retry on the requests and ensure we get N succesful requests in a row (to not pass on something flaky). The echo.Call does this all automatically

Reply:
Unfortunately it's not so simple in this case. I observed false-positive results without sleep. While iterating over test cases, only traffic configurations are changed, so it's possible that test `n` succeeds, because it was executed before its configuration was pushed to istio-proxy and therefore it used configuration of test `n-1`. To avoid this problem I added sleep after deleting Istio resources to make sure that the next test case will run with proper configuration.